routes/api/project/location/vm.rb

```diff
@@ -5,50 +5,86 @@ class CloverApi
+        next_cursor: result[:next_cursor],
+        count: result[:count]
+      }
     end
 
-    r.post true do
-      Authorization.authorize(@current_user.id, "Vm:create", @project.id)
-      fail Validation::ValidationFailed.new({billing_info: "Project doesn't have valid billing information"}) unless @project.has_valid_payment_method?
-
-      st = Prog::Vm::Nexus.assemble(
-        r.params["public_key"],
-        @project.id,
-        name: r.params["name"],
-        unix_user: r.params["unix_user"],
-        size: r.params["size"],
-        location: @location,
-        boot_image: r.params["boot_image"],
-        enable_ip4: !!r.params["enable_ip4"]
-      )
+    r.on "ubid" do
+      r.is String do |vm_ubid|
+        vm = Vm.from_ubid(vm_ubid)
+
+        r.get true do
+          get_vm_handler(@current_user, vm)
+        end
 
-      serialize(st.subject)
+        r.delete true do
+          delete_vm_handler(@current_user, vm)
+        end
+      end
     end
 
     r.is String do |vm_name|
-      vm = @project.vms_dataset.where(location: @location).where { {Sequel[:vm][:name] => vm_name} }.first
+      r.post true do
+        Authorization.authorize(@current_user.id, "Vm:create", @project.id)
+        fail Validation::ValidationFailed.new({billing_info: "Project doesn't have valid billing information"}) unless @project.has_valid_payment_method?
+
+        required_parameters = ["public_key"]
+        allowed_optional_parameters = ["size", "unix_user", "location", "boot_image", "enable_ip4"]
 
-      unless vm
-        response.status = 404
-        r.halt
+        request_body = r.body.read
+        Validation.validate_request_body(request_body, required_parameters, allowed_optional_parameters)
+
+        request_body_params = JSON.parse(request_body)
+
+        st = Prog::Vm::Nexus.assemble(
+          request_body_params["public_key"],
+          @project.id,
+          name: vm_name,
+          **request_body_params.except(*required_parameters).transform_keys { |key| key.to_sym }
+        )
+
+        serialize(st.subject)
       end
 
-      r.get true do
-        Authorization.authorize(@current_user.id, "Vm:view", vm.id)
+      vm = @project.vms_dataset.where(location: @location).where { {Sequel[:vm][:name] => vm_name} }.first
 
-        serialize(vm)
+      r.get true do
+        get_vm_handler(@current_user, vm)
       end
 
       r.delete true do
-        Authorization.authorize(@current_user.id, "Vm:delete", vm.id)
+        delete_vm_handler(@current_user, vm)
+      end
+    end
+  end
+
+  def get_vm_handler(user, vm)
+    unless vm
+      response.status = 404
+      request.halt
+    end
 
-        vm.incr_destroy
+    Authorization.authorize(user.id, "Vm:view", vm.id)
 
-        serialize(vm)
-      end
+    serialize(vm)
+  end
+
+  def delete_vm_handler(user, vm)
+    if vm
+      Authorization.authorize(user.id, "Vm:delete", vm.id)
+      vm.incr_destroy
     end
+
+    response.status = 204
+    request.halt
```

Comment:
I don't understand, why do we halt?

Reply:
To complete processing request at that point, otherwise roda won't be able to return properly.